torch/jit.py

```diff
@@ -0,0 +1,214 @@
+import torch.autograd.function as function
+import torch._C
+from torch.autograd import Variable
+from torch.nn import Module
+import itertools
+import types
+import contextlib
+import os
+# Example how to use:
+#
+# import torch.jit
+# model = model.RNNModel(args.model, ...)
+# model = torch.jit.traced(model)
+
+
+def flatten(x):
+    """
+    Flatten an arbitrarily nested structure of Variables into
+    a tuple of Variables.
+    """
+    return tuple(function._iter_variables(x))
+
+
+@contextlib.contextmanager
+def _fork_rng(enabled=True):
+    """
+    Forks the RNG, so that when you return, the RNG is reset
+    to the state that it was previously in.  This is important
+    if we are evaluating a trace twice, and it incorporates
+    randomness: if we don't reset the seed, we might get totally
+    different results!
+
+    TODO: Randomness in models is a big problem for reproduceability,
+    because it means if we start executing models out of order,
+    they may behave differently.  Interesting question is whether
+    or not backwards pass ever has randomness.  I hope not.
+    """
+    if not enabled:
+        yield
+        return
+
+    cpu_rng_state = torch.get_rng_state()
+    gpu_rng_state = None
+    if torch.cuda.is_available():
+        gpu_rng_state = torch.cuda.get_rng_state()
+
+    yield
+
+    torch.set_rng_state(cpu_rng_state)
+    if gpu_rng_state is not None:
+        torch.cuda.set_rng_state(gpu_rng_state)
```

Comment:
I think this only caches the RNG state of the current GPU.

Reply:
Indeed, this seems to be the case. Should we add a torch.cuda helper for setting all of the states? (Maybe set_rng_state should set everything, and there should be a set_current_gpu_rng_state...)